apps/prairielearn/src/middlewares/authzCourseOrInstance.ts

```diff
@@ -205,420 +431,272 @@ export async function authzCourseOrInstance(req: Request, res: Response) {
-    if (!isValid(req_date)) {
-      debug(`requested date is invalid: ${req.cookies.pl2_requested_date}, ${req_date}`);
+    const result = await getOverrideUserData({
+      requestedUid: req.cookies.pl2_requested_uid,
+      isAdministrator: res.locals.is_administrator,
+      courseInstanceId: req.params.course_instance_id || null,
+    });
+    if (!result.success) {
       clearOverrideCookies(res, overrides);
-
-      throw new AugmentedError('Access denied', {
-        status: 403,
-        info: html`
-          <p>
-            You have requested an invalid effective date:
-            <code>${req.cookies.pl2_requested_date}</code>. All requested changes to the effective
-            user have been removed.
-          </p>
-        `,
-      });
+      throw result.error;
     }
-
-    debug(`effective req_date = ${req_date}`);
-  }
-
-  const effectiveAuthzData = await selectAuthzData({
-    user_id: user.user_id,
-    course_id: req.params.course_id || null,
-    course_instance_id: req.params.course_instance_id || null,
-    is_administrator,
-    allow_example_course_override: false,
-    ip: req.ip || null,
-    req_date,
-    req_mode: res.locals.authz_data.mode,
-    req_course_role: req.cookies.pl2_requested_course_role || null,
-    req_course_instance_role: req.cookies.pl2_requested_course_instance_role || null,
+    return result.value;
   });
 
-  // If the authn user were denied access, then we would return an error. Here,
-  // we simply return (without error). This allows the authn user to keep access
-  // to pages (e.g., the effective user page) for which only authn permissions
-  // are required.
-  if (effectiveAuthzData === null) {
-    debug('effective user was denied access');
-
-    res.locals.authz_data.user = user;
-    res.locals.authz_data.is_administrator = false;
-
-    res.locals.authz_data.course_role = 'None';
-    res.locals.authz_data.has_course_permission_preview = false;
-    res.locals.authz_data.has_course_permission_view = false;
-    res.locals.authz_data.has_course_permission_edit = false;
-    res.locals.authz_data.has_course_permission_own = false;
-
-    if (req.params.course_instance_id) {
-      res.locals.authz_data.course_instance_role = 'None';
-      res.locals.authz_data.has_course_instance_permission_view = false;
-      res.locals.authz_data.has_course_instance_permission_edit = false;
-      res.locals.authz_data.has_student_access = false;
-      res.locals.authz_data.has_student_access_with_enrollment = false;
-
-      if (res.locals.authz_data.user.uid !== res.locals.authz_data.authn_user.uid) {
-        res.locals.authz_data.user_with_requested_uid_has_instructor_access_to_course_instance =
-          user_with_requested_uid_has_instructor_access_to_course_instance;
-      }
+  const {
+    authzData: effectiveAuthResult,
+    course: effectiveCourse,
+    institution: effectiveInstitution,
+    courseInstance: effectiveCourseInstance,
```

Comment:
👀 when would we expect these to ever differ from the non-effective-permissions results of `constructCourseOrInstanceContext`? Put differently, I can't tell why the caller cares about these.

Reply:
Fixed in 34aa61a826142203e468da77ab516ce25d525421.